[BUG] test_window_aggregate_udf_array_from_python fails on databricks

**Describe the bug**
we started to run more tests on databricks and we found that the udf_test.py test_window_aggregate_udf_array_from_python is failing because the script is running with enabling assertions.  Its a bit weird it worked without the assertions enabled as I would have thought the output would be different but maybe something is allow to work.  We need to investigate what is going on. Seems like a schema not matching column types issue:
```
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[No_Partition-Byte][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[No_Partition-Short][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[No_Partition-Integer][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded-Byte][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded-Short][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded-Integer][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded_Following-Byte][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded_Following-Short][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded_Following-Integer][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded_Preceding-Byte][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded_Preceding-Short][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Unbounded_Preceding-Integer][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Lower_Upper-Byte][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Lower_Upper-Short][IGNORE_ORDER]
FAILED src/main/python/udf_test.py::test_window_aggregate_udf_array_from_python[Lower_Upper-Integer][IGNORE_ORDER]
```

```
java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=11, type=LIST, nullCount=Optional.empty, offHeap=(ID: 521 563d5d2774f0)}], cudfTable=94821486278400, rows=11} to [LongType]
	at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:569)
	at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:455)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.$anonfun$read$2(GpuArrowEvalPythonExec.scala:357)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at org.apache.spark.sql.rapids.execution.python.GpuArrowPythonRunner.withResource(GpuArrowEvalPythonExec.scala:388)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.read(GpuArrowEvalPythonExec.scala:354)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.read(GpuArrowEvalPythonExec.scala:367)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.read(GpuArrowEvalPythonExec.scala:318)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:551)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.next(PythonRunner.scala:559)
	at org.apache.spark.InterruptibleIterator.next(InterruptibleIterator.scala:40)
	at org.apache.spark.sql.rapids.execution.python.GpuWindowInPandasExecBase$$anon$1.next(GpuWindowInPandasExecBase.scala:565)
	at org.apache.spark.sql.rapids.execution.python.GpuWindowInPandasExecBase$$anon$1.next(GpuWindowInPandasExecBase.scala:540)
	at com.nvidia.spark.rapids.GpuRangePartitioner$$anon$1.loadNextBatch(GpuRangePartitioner.scala:180)
	at com.nvidia.spark.rapids.GpuRangePartitioner$$anon$1.hasNext(GpuRangePartitioner.scala:194)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at com.nvidia.spark.rapids.SamplingUtils$.reservoirSampleAndCount(SamplingUtils.scala:46)
	at com.nvidia.spark.rapids.GpuRangePartitioner.$anonfun$sketch$1(GpuRangePartitioner.scala:49)
	at com.nvidia.spark.rapids.GpuRangePartitioner.$anonfun$sketch$1$adapted(GpuRangePartitioner.scala:46)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsWithIndex$2(RDD.scala:896)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsWithIndex$2$adapted(RDD.scala:896)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:60)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:356)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:320)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:144)
	at org.apache.spark.scheduler.Task.run(Task.scala:117)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$9(Executor.scala:640)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1581)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:643)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
21/02/01 21:43:51 WARN TaskSetManager: Lost task 0.0 in stage 60.0 (TID 4230, ip-10-59-236-6.us-west-2.compute.internal, executor driver): java.lang.AssertionError: Type conversion is not allowed from Table{columns=[ColumnVector{rows=11, type=LIST, nullCount=Optional.empty, offHeap=(ID: 521 563d5d2774f0)}], cudfTable=94821486278400, rows=11} to [LongType]
	at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:569)
	at com.nvidia.spark.rapids.GpuColumnVector.from(GpuColumnVector.java:455)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.$anonfun$read$2(GpuArrowEvalPythonExec.scala:357)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at org.apache.spark.sql.rapids.execution.python.GpuArrowPythonRunner.withResource(GpuArrowEvalPythonExec.scala:388)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.read(GpuArrowEvalPythonExec.scala:354)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.read(GpuArrowEvalPythonExec.scala:367)
	at org.apache.spark.sql.rapids.execution.python.GpuPythonArrowOutput$$anon$1.read(GpuArrowEvalPythonExec.scala:318)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:551)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.next(PythonRunner.scala:559)
	at org.apache.spark.InterruptibleIterator.next(InterruptibleIterator.scala:40)
	at org.apache.spark.sql.rapids.execution.python.GpuWindowInPandasExecBase$$anon$1.next(GpuWindowInPandasExecBase.scala:565)
	at org.apache.spark.sql.rapids.execution.python.GpuWindowInPandasExecBase$$anon$1.next(GpuWindowInPandasExecBase.scala:540)
	at com.nvidia.spark.rapids.GpuRangePartitioner$$anon$1.loadNextBatch(GpuRangePartitioner.scala:180)
	at com.nvidia.spark.rapids.GpuRangePartitioner$$anon$1.hasNext(GpuRangePartitioner.scala:194)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at com.nvidia.spark.rapids.SamplingUtils$.reservoirSampleAndCount(SamplingUtils.scala:46)
	at com.nvidia.spark.rapids.GpuRangePartitioner.$anonfun$sketch$1(GpuRangePartitioner.scala:49)
	at com.nvidia.spark.rapids.GpuRangePartitioner.$anonfun$sketch$1$adapted(GpuRangePartitioner.scala:46)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsWithIndex$2(RDD.scala:896)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsWithIndex$2$adapted(RDD.scala:896)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:60)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:356)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:320)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:144)
	at org.apache.spark.scheduler.Task.run(Task.scala:117)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$9(Executor.scala:640)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1581)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:643)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

--- a/integration_tests/src/main/python/udf_test.py
+++ b/integration_tests/src/main/python/udf_test.py
@@ -14,7 +14,7 @@
 
 import pytest
 
-from conftest import is_at_least_precommit_run, is_databricks_runtime
+from conftest import is_at_least_precommit_run
 
 from pyspark.sql.pandas.utils import require_minimum_pyarrow_version, require_minimum_pandas_version
 try:
@@ -170,8 +170,6 @@ def test_window_aggregate_udf(data_gen, window):
         conf=arrow_udf_conf)
 
 
-@pytest.mark.xfail(condition=is_databricks_runtime(),
-        reason='https://github.com/NVIDIA/spark-rapids/issues/1644')
 @ignore_order
 @pytest.mark.parametrize('data_gen', [byte_gen, short_gen, int_gen], ids=idfn)
 @pytest.mark.parametrize('window', udf_windows, ids=window_ids)

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala
@@ -555,6 +555,11 @@ case class GpuArrowEvalPythonExec(
 
     // cache in a local to avoid serializing the plan
     val inputSchema = child.output.toStructType
+    // Build the Python output schema from UDF expressions instead of the 'resultAttrs', because
+    // the 'resultAttrs' is NOT always equal to the Python output schema. For example,
+    // On Databricks when projecting only one column from a Python UDF output where containing
+    // multiple result columns, there will be only one attribute in the 'resultAttrs' for the
+    // projecting output, but the output schema for this Python UDF contains multiple columns.
     val pythonOutputSchema = StructType.fromAttributes(udfs.map(_.resultAttribute))
 
     val childOutput = child.output

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala
@@ -555,7 +555,7 @@ case class GpuArrowEvalPythonExec(
 
     // cache in a local to avoid serializing the plan
     val inputSchema = child.output.toStructType
-    val pythonOutputSchema = StructType.fromAttributes(resultAttrs)
+    val pythonOutputSchema = StructType.fromAttributes(udfs.map(_.resultAttribute))
 
     val childOutput = child.output
     val targetBatchSize = batchSize

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala
@@ -490,6 +490,13 @@ trait GpuWindowInPandasExecBase extends UnaryExecNode with GpuExec {
 
     lazy val isPythonOnGpuEnabled = GpuPythonHelper.isPythonOnGpuEnabled(conf, pythonModuleKey)
     // cache in a local to avoid serializing the plan
+
+    // Build the Python output schema from UDF expressions instead of the 'windowExpression',
+    // because the 'windowExpression' does NOT always represent the Python output schema.
+    // For example, on Databricks when projecting only one column from a Python UDF output
+    // where containing multiple result columns, there will be only one item in the
+    // 'windowExpression' for the projecting output, but the output schema for this Python
+    // UDF contains multiple columns.
     val pythonOutputSchema = StructType.fromAttributes(udfExpressions.map(_.resultAttribute))
     val childOutput = child.output
 

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala
@@ -490,8 +490,7 @@ trait GpuWindowInPandasExecBase extends UnaryExecNode with GpuExec {
 
     lazy val isPythonOnGpuEnabled = GpuPythonHelper.isPythonOnGpuEnabled(conf, pythonModuleKey)
     // cache in a local to avoid serializing the plan
-    val retAttributes = windowExpression.map(_.asInstanceOf[NamedExpression].toAttribute)
-    val pythonOutputSchema = StructType.fromAttributes(retAttributes)
+    val pythonOutputSchema = StructType.fromAttributes(udfExpressions.map(_.resultAttribute))
     val childOutput = child.output
 
     // 8) Start processing.

